netty/src/main/java/io/grpc/netty/ProtocolNegotiator.java

```diff
@@ -51,16 +51,16 @@
      * The HTTP/2 scheme to be used when sending {@code HEADERS}.
      */
     AsciiString scheme();
+
+    /**
+     * Returns a set of attributes, which varies depending on the handler implementation.
+     */
+    Attributes getAttributes();
```

Comment:
even it is called too early, we can still return an attribute like TosProvider, which can give valid TOS value at sometime later when the connection is established. Using `handleProtocolNegotiationCompleted` callback is not so easy since its impl (in `NettyClientHandler` (I'm only talking about client side)) is hard to access the transport or transport attribute.